### DISCUSSION TITLE
Settings and default values for general-purpose use

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#498] [Q&A] [answered]
I'm in the process of exposing some of Jolt's settings as "project settings" through my Godot extension, which means that they'll be configurable through the Godot editor and any overridden values are stored within the user's Godot project, as opposed to just being hardcoded on my side. This involves setting default values for them, which has got me wondering what kind of defaults I should be aiming for and what the potential consequences of those defaults could be.

I picked out a couple of settings that I thought seemed somewhat essential or useful. Now, I am not a physics developer, nor someone who typically interacts much with one, so this is very much a layman's take on things, but this is roughly the list of settings that Godot developers would be able to modify:

1. Number of solver velocity iterations (from `PhysicsSettings`)
1. Number of solver position iterations (from `PhysicsSettings`)
1. Baumgarte stabilization factor (from `PhysicsSettings`)
1. Allow sleeping (from `PhysicsSettings`)
1. Time before sleep (from `PhysicsSettings`)
1. Point velocity sleep threshold (from `PhysicsSettings`)
1. Maximum number of bodies (from `PhysicsSystem::Init`)
1. Maximum number of body pairs (from `PhysicsSystem::Init`)
1. Maximum number of in-flight body pairs (from `PhysicsSettings`)
1. Maximum number of contact constraints (from `PhysicsSystem::Init`)
1. Maximum number of allocated jobs (self-imposed, but very similar to `JobSystemThreadPool`)
1. Maximum number of job barriers (from `JobSystemWithBarrier`)
1. Temporary memory size (self-imposed, but very similar to `TempAllocatorImpl`)

I realize this is a very "it depends" kind of question, but would you say that's a good set of basic/essential/useful settings to start with, or am I leaving out something crucial that users/integrators typically tinker with? If any of you watchers are reading this I'd love to hear what settings you might have exposed or maybe even changed.

As for default values, I'm not sure how conservative I should be with the above mentioned maximums. When reading the comments in `HelloWorld` it's made to seem like I should crank up most of them quite high, at least compared to what `HelloWorld` uses. When considering that most users probably won't pay attention to these settings, much less modify them, can I comfortably use those "high" values and not feel like I'm just wasting people's memory?

If I were to go a more conservative route, and set the default values to something like what `HelloWorld` uses, what kind of errors can I expect from Jolt when users inevitably bump up against these limits? Is it largely considered undefined behavior once you do, or can I reasonably detect this and emit some kind of user-friendly error about what limit was hit?

For example, I noticed that `BodyInterface::CreateBody` returns `nullptr` when going past the maximum number of bodies, which is easy enough to report on, but when going past the maximum number of contact constraints I instead get asserts and bodies start falling through the world. This not only necessitates that I enable asserts, but also that I forward them to Godot and hope that the user can infer what's wrong, if not plainly stated in the assert.

That ended up being a long set of question, but I appreciate any input. :)

**Comment by jrouwe** [comment, marked as answer]
Hello,

I would add PhysicsSettings::mSpeculativeContactDistance and mPenetrationSlop.

I would remove mMaxInFlightBodyPairs as the default should be plenty for everyone.

It is probably better to take the max bodies/constraints and body pairs from the SampleApp because the HelloWorld app uses a really low value:

```
static constexpr uint cNumBodies = 10240;
static constexpr uint cMaxBodyPairs = 65536;
static constexpr uint cMaxContactConstraints = 20480;
```

But they should indeed be user configurable.

For max allocated jobs and barriers I would just use ```cMaxPhysicsJobs``` and ```cMaxPhysicsBarriers```. The number of jobs doesn't really depend on the scene so these values should be good.

The default temp memory size you probably also want to take from the sample app (32M) but should definitively be user configurable

> I instead get asserts and bodies start falling through the world. This not only necessitates that I enable asserts, but also that I forward them to Godot and hope that the user can infer what's wrong, if not plainly stated in the assert.

I should probably make a listener of some sorts for this instead of triggering an assert. 

Regards,

Jorrit.

**Reply by mihe** [reply]
Great, thank you! I'll go ahead and implement the changes you suggested.

> I should probably make a listener of some sorts for this instead of triggering an assert.

That would be very helpful.

**Reply by jrouwe** [reply]
Added in 45f147f70008c57be92f561e1a5b2ff18f7e01e9

**Reply by mihe** [reply]
Fantastic, I'll give it a try shortly, thank you!

**Reply by mihe** [reply]
Seeing as how Godot doesn't really have any notion of a maximum linear/angular velocity on bodies, would you say it makes sense to expose `BodyCreationSettings::mMaxLinearVelocity` and `BodyCreationSettings::mMaxAngularVelocity` as project-wide settings as well? Are there any potential side-effects from setting these too high?

(`EPhysicsUpdateError` works great by the way, thanks again!)

**Reply by jrouwe** [reply]
I think that would be fine. That limit is mostly a last resort to prevent the system from exploding in case something keeps adding energy to the system, so you probably don't want to set it much higher than the current default.
